File: app/src/main/java/org/bitcoindotcom/bchprocessor/bip70/model/InvoiceStatus.java
package org.bitcoindotcom.bchprocessor.bip70.model;

import com.bitcoin.merchant.app.util.GsonUtil;
import com.google.gson.annotations.SerializedName;

import java.util.Arrays;
import java.util.Date;

public class InvoiceStatus {
    @SerializedName("paymentUrl")
    public String paymentUrl;
    @SerializedName("paymentId")
    public String paymentId;
    @SerializedName("memo")
    public String memo;
    @SerializedName("time")
    public Date time; // 2019-10-05T21:44:34.009Z
    @SerializedName("expires")
    public Date expires; // 2019-10-05T21:59:34.009Z
    @SerializedName("status")
    public String status;
    @SerializedName("fiatSymbol")
    public String fiatSymbol; // "USD"
    @SerializedName("fiatRate")
    public double fiatRate; // 222.67
    @SerializedName("txId")
    public String txId;
    @SerializedName("merchantId")
    public String merchantId;
    @SerializedName("outputs")
    public InvoiceStatusOutput[] outputs = new InvoiceStatusOutput[0];

    public static InvoiceStatus fromJson(String message) {
        return GsonUtil.INSTANCE.getGson().fromJson(message, InvoiceStatus.class);
    }

    public String getFirstAddress() {
        return outputs.length > 0 ? outputs[0].address : null;
    }

    public long getTotalBchAmount() {
        long amount = 0;
        for (InvoiceStatusOutput output : outputs) {
            amount += output.amount;
        }
        return amount;
    }

    public boolean isPaid() {
        return Status.paid.name().equals(status);
    }

    public boolean isOpen() {
        return Status.open.name().equals(status);
    }

    public boolean isExpired() {
        return Status.expired.name().equals(status);
    }

    public String getWalletUri() {
        return "bitcoincash:?r=" + paymentUrl;
    }

    @Override
    public String toString() {
        return "InvoiceStatus{" +
                "paymentUrl='" + paymentUrl + '\'' +
                ", paymentId='" + paymentId + '\'' +
                ", memo='" + memo + '\'' +
                ", time='" + time + '\'' +
                ", expires='" + expires + '\'' +
                ", status='" + status + '\'' +
                ", txId='" + txId + '\'' +
                ", outputs=" + Arrays.toString(outputs) +
                '}';
    }

    enum Status {
        open, paid, expired
    }
}

File: app/src/main/java/org/bitcoindotcom/bchprocessor/bip70/model/InvoiceStatus.kt
package org.bitcoindotcom.bchprocessor.bip70.model

import com.google.gson.annotations.SerializedName
import org.bitcoindotcom.bchprocessor.bip70.GsonHelper.gson
import java.util.*
import kotlin.collections.ArrayList

data class InvoiceStatus @JvmOverloads constructor(
        @SerializedName("paymentUrl") var paymentUrl: String = "",
        @SerializedName("paymentId") var paymentId: String = "",
        @SerializedName("paymentAsset") var paymentAsset: String? = "",
        @SerializedName("memo") var memo: String? = null,
        @SerializedName("time") var time: Date = Date(),
        @SerializedName("expires") var expires: Date = Date(),
        @SerializedName("status") var status: String = Status.expired.name,
        @SerializedName("network") var network: String? = "",
        @SerializedName("currency") var currency: String? = "",
        @SerializedName("fiatSymbol") var fiatSymbol: String? = "",
        @SerializedName("fiatRate") var fiatRate: Double = 0.0,
        @SerializedName("fiatTotal") var fiatTotal: Double = 0.0,
        @SerializedName("txId") var txId: String? = null,
        @SerializedName("merchantId") var merchantId: String? = null,
        @SerializedName("outputs") var outputs: List<InvoiceStatusOutput> = ArrayList(0)) {

    val isInitialized: Boolean
        get() = paymentUrl.isNotEmpty() && paymentId.isNotEmpty() && fiatTotal != 0.0 && outputs.isNotEmpty()
    val firstAddress: String?
        get() = if (outputs.isNotEmpty()) outputs[0].address else null

    val totalBchAmount: Long
        get() {
            var amount: Long = 0
            for (output in outputs) {
                amount += output.amount
            }
            return amount
        }

    val isPaid: Boolean
        get() = Status.paid.name == status

    val isOpen: Boolean
        get() = Status.open.name == status

    val isExpired: Boolean
        get() = Status.expired.name == status

    val walletUri: String
        get() = "bitcoincash:?r=$paymentUrl"

    internal enum class Status {
        open, paid, expired
    }

    companion object {
        fun fromJson(message: String?): InvoiceStatus {
            return gson.fromJson(message, InvoiceStatus::class.java)
        }
    }
}